### PR TITLE
commands/.../new,pkg/scaffold/helm: source chart flag for new projects

### PR DESCRIPTION
**Pull request by joelanford**
**Description of the change:**
This PR adds new flags to the `operator-sdk new` command that enable new Helm projects to be created from existing Helm charts.
* `--helm-chart` - The chart to use when creating the project. Format is:
  * `<url>`
  * `<repo>/<chart_name>`
  * `<chart_name>` (with `--helm-chart-repo` specified)
  * `/path/to/chart.tgz` (local chart archive)
  * `/path/to/chartDir` (local chart directory)
* `--helm-chart-version` - The version of the chart to fetch from a remote repository.
* `--helm-chart-repo` - The URL of the Helm chart repo in which to find the chart specified by `--helm-chart`

It also makes `--api-version` and `--kind` optional if `--helm-chart` is set, If left unset, `--api-version` is defaulted to `charts.helm.k8s.io/v1alpha1` and `--kind` is deduced from the fetched chart.

Examples:
* From a URL:
    ```
    operator-sdk new memcached-operator --type=helm --helm-chart=https://storage.googleapis.com/kubernetes-charts/memcached-2.4.0.tgz
    ```

* From stable repo:
    ```
    operator-sdk new memcached-operator --type=helm --helm-chart=stable/memcached
    ```

* From custom repo:
    ```
    operator-sdk new memcached-operator --type=helm --helm-chart=memcached --helm-chart-repo=https://storage.googleapis.com/kubernetes-charts
    ```
* From local archive: 
    ```
    operator-sdk new memcached-operator --type=helm --helm-chart=/path/to/memcached-2.4.0.tgz
    ```

* From a local directory:
    ```
    operator-sdk new memcached-operator --type=helm --helm-chart=/path/to/memcached/
    ```

**Motivation for the change:**
At the suggestion of @robszumski, I put together this PR to simplify the UX for building operators from existing Helm charts.

**NOTE**: This PR exposes only some of the flags that `helm fetch` does, so not all use cases are supported (e.g. users will not be able to authenticate to a remote repo using the operator-sdk CLI from this PR).

### CODE REVIEW

File: commands/operator-sdk/cmd/new.go [hunk 1]
```diff
@@ -56,6 +56,10 @@ generates a skeletal app-operator application in $GOPATH/src/github.com/example.
 	newCmd.Flags().BoolVar(&generatePlaybook, "generate-playbook", false, "Generate a playbook skeleton. (Only used for --type ansible)")
 	newCmd.Flags().BoolVar(&isClusterScoped, "cluster-scoped", false, "Generate cluster-scoped resources instead of namespace-scoped")
 
+	newCmd.Flags().StringVar(&helmChartRef, "helm-chart", "", "Initialize helm operator with specified chart (<URL>, <repo>/<name>, or local path")
```
Comment (LiliC) [inline commands/operator-sdk/cmd/new.go]: Suggestion:
```suggestion
	newCmd.Flags().StringVar(&helmChartRef, "helm-chart", "", "Initialize helm operator with existing helm chart (<URL>, <repo>/<name>, or local path")
```

File: commands/operator-sdk/cmd/new.go [hunk 2]
```diff
@@ -300,11 +312,24 @@ func verifyFlags() error {
...
+			log.Fatal("Value of --helm-chart-version can only be used with --type `helm` and --helm-chart")
+		}
+	}
+
 	if operatorType == projutil.OperatorTypeGo && (len(apiVersion) != 0 || len(kind) != 0) {
 		return fmt.Errorf("operators of type Go do not use --api-version or --kind")
 	}
 
-	if operatorType != projutil.OperatorTypeGo {
+	if operatorType == projutil.OperatorTypeAnsible || operatorType == projutil.OperatorTypeHelm && len(helmChartRef) == 0 {
```
Comment (LiliC) [inline commands/operator-sdk/cmd/new.go]: Should we add a quick comment here to avoid confusion in the future as to why this is so specific?

Something like this should be good:

> It also makes --api-version and --kind optional if --helm-chart is set, If left unset, --api-version is defaulted to charts.helm.k8s.io/v1alpha1 and --kind is deduced from the fetched chart.

File: pkg/scaffold/helm/chart.go [hunk 3]
```diff
@@ -15,40 +15,185 @@
...
+	DefaultAPIVersion string = "charts.helm.k8s.io/v1alpha1"
+)
+
+type FetchChartOptions struct {
+	Chart   string
+	Repo    string
+	Version string
+}
+
+func CreateChart(apiVersion, kind string, opts FetchChartOptions, projectDir string) (*scaffold.Resource, *chart.Chart, error) {
```
Comment (LiliC) [inline pkg/scaffold/helm/chart.go]: Missing function comment above.

File: pkg/scaffold/helm/chart.go [hunk 4]
```diff
@@ -15,40 +15,185 @@
...
+	// from Helm's default template. Otherwise, fetch it.
+	if len(opts.Chart) == 0 {
+		r, c, err = scaffoldChart(apiVersion, kind, chartsDir)
+	} else {
+		r, c, err = fetchChart(apiVersion, kind, opts, chartsDir)
+	}
+	if err != nil {
+		return nil, nil, err
+	}
+	log.Infof("Create %s/%s/", HelmChartsDir, c.GetMetadata().GetName())
```
Comment (LiliC) [inline pkg/scaffold/helm/chart.go]: To me the log implies the user should create this.

File: pkg/scaffold/helm/chart.go [hunk 5]
```diff
@@ -15,40 +15,185 @@
...
+
+func scaffoldChart(apiVersion, kind, destDir string) (*scaffold.Resource, *chart.Chart, error) {
+	r, err := scaffold.NewResource(apiVersion, kind)
+	if err != nil {
+		return nil, nil, err
+	}
 
 	chartfile := &chart.Metadata{
-		Name:        r.LowerKind,
+		Name:        strcase.ToDelimited(kind, '-'),
```
Comment (LiliC) [inline pkg/scaffold/helm/chart.go]: Just curious to know why this change.

File: pkg/scaffold/helm/chart.go (hunk 4 above)
Comment (joelanford) [reply pkg/scaffold/helm/chart.go]: I don't disagree. However, this log matches other logs when scaffolding files in a new project, so I went with that to be consistent.

https://github.com/operator-framework/operator-sdk/blob/5f3c3a50fa121e6cae05b9afced1469bc9fe2b7c/pkg/scaffold/scaffold.go#L171

File: pkg/scaffold/helm/chart.go (hunk 5 above)
Comment (joelanford) [reply pkg/scaffold/helm/chart.go]: Many of the multi-word chart names use the hyphenated format: https://github.com/helm/charts/tree/master/stable

However, I'm looking at the [helm chart naming convention](https://github.com/helm/helm/blob/master/docs/chart_best_practices/conventions.md#chart-names), which says hyphens are allowed. But based on some of the problems mentioned in the [linked issue](https://github.com/helm/helm/issues/2192), it seems like it would be better to just avoid all of that. I'll revert back to `r.LowerKind`.

File: pkg/scaffold/helm/chart.go (hunk 4 above)
Comment (joelanford) [reply pkg/scaffold/helm/chart.go]: Yep. I think it makes sense. It looks like "Created" would make sense here and in scaffold.go.

File: pkg/scaffold/helm/chart.go [hunk 6]
```diff
@@ -15,40 +15,234 @@
...
+		}
+		opts.Chart = chartURL
+	}
+
+	tmpDir, err := ioutil.TempDir("", "osdk-helm-chart")
+	if err != nil {
 		return nil, err
 	}
-	chartDir, err := chartutil.Create(chartfile, chartsDir)
+	defer os.RemoveAll(tmpDir)
```
Comment (estroz) [inline pkg/scaffold/helm/chart.go]: Log this error.

**Comment by hasbro17** [comment]
@joelanford We can probably mention this feature in the CHANGELOG.

Also the godocs for `CreateChart` are pretty detailed 👍 but we might still want to update the CLI reference and user-guide to showcase the `--helm-chart` flag. It's fine if we do a follow up based on this PR. Just wanted to point it out.

File: commands/operator-sdk/cmd/new.go [hunk 7]
```diff
@@ -300,11 +314,28 @@ func verifyFlags() error {
 	if operatorType != projutil.OperatorTypeAnsible && generatePlaybook {
 		return fmt.Errorf("value of --generate-playbook can only be used with --type `ansible`")
 	}
+
+	if operatorType != projutil.OperatorTypeHelm {
+		if len(helmChartRef) != 0 {
+			return fmt.Errorf("value of --helm-chart can only be used with --type `helm`")
+		}
+		if len(helmChartRef) == 0 && len(helmChartRepo) != 0 {
```
[the remark this answers is not in the document]
Comment (joelanford) [reply commands/operator-sdk/cmd/new.go]: Thanks. There was actually more wrong with this logic in general, so I refactored it.

**Comment by joelanford** [comment]
@hasbro Thanks! I'll go ahead and update the CLI doc so I can reference it in the CHANGELOG.

File: pkg/scaffold/helm/watches.go [hunk 8]
```diff
@@ -36,12 +37,15 @@ func (s *WatchesYAML) GetInput() (input.Input, error) {
...
+	}
 	return s.Input, nil
 }
 
 const watchesYAMLTmpl = `---
 - version: {{.Resource.Version}}
   group: {{.Resource.FullGroup}}
   kind: {{.Resource.Kind}}
-  chart: /opt/helm/{{.HelmChartsDir}}/{{.Resource.LowerKind}}
+  chart: /opt/helm/{{.HelmChartsDir}}/{{.ChartName}}
```
[the remark this answers is not in the document]
Comment (joelanford) [reply pkg/scaffold/helm/watches.go]: According to the [chart name conventions](https://github.com/helm/helm/blob/master/docs/chart_best_practices/conventions.md#chart-names), chart names SHOULD be lower case. However, the directory name MUST match the name in Chart.yaml. It seems the chart name case is not enforced, but the directory matching  the name in Chart.yaml is.

```console
$ helm create Test
Creating Test

$ helm package Test
Successfully packaged chart and saved it to: /tmp/Test-0.1.0.tgz

$ mv Test test

$ helm package test
Error: directory name (test) and Chart.yaml name (Test) must match
```

Since our code works by using Helm's packages to load the chart metadata, `.ChartName` is effectively a read-only field provided by Helm. For that reason, I think I lean toward accepting what Helm gives us without modification. I would expect Helm to error out if a user tried to create or fetch an invalid chart.
